### AINAHACK - 2023_centres_civics_activitats_formatives.csv

### IMPORTAR PAQUETES

In [21]:
import numpy as np
import pandas as pd
from janitor import clean_names

# Automcompletar rápido
%config IPCompleter.greedy=True

### IMPORTAR LOS DATOS

Sustituir el nombre_fichero_datos.

In [22]:
nombre_fichero_datos = '2023_centres_civics_activitats_formacio.csv'

Cargar los datos.

In [23]:
# Fix the path to use proper directory separators and a more robust path construction
import os

# Update the project path with the complete path
ruta_proyecto = os.path.join(os.path.expanduser("~"), "Downloads", "YTT", "AINAHACK", "AinaHack-Data")

# Update the path to match the actual file location
ruta_completa = os.path.join(ruta_proyecto, "02_Datos", "01_Originales", nombre_fichero_datos)

# Add error handling for file reading
try:
    df = pd.read_csv(ruta_completa, encoding='utf-8')
    print(f"Archivo cargado exitosamente desde: {ruta_completa}")
except FileNotFoundError:
    print(f"Error: No se encontró el archivo en la ruta: {ruta_completa}")
    print("Por favor, verifica que:")
    print("1. La ruta del proyecto es correcta:", ruta_proyecto)
    print("2. El archivo existe en la carpeta 02_Datos/01_Originales")
    print("3. El nombre del archivo es exactamente '2023_centres_civics_activitats_formacio.csv'")

df = pd.read_csv(ruta_completa,index_col=0)
df

Archivo cargado exitosamente desde: C:\Users\Usuario\Downloads\YTT\AINAHACK\AinaHack-Data\02_Datos\01_Originales\2023_centres_civics_activitats_formacio.csv


Equipament                       Indicador  \
Any                                                                          
2023              Centre Cívic Barceloneta            ActivitatsDeFormacio   
2023              Centre Cívic Barceloneta  InscripcionsActivitatsFormacio   
2023              Centre Cívic Barceloneta                 Metres_quadrats   
2023            Centre Cívic Baró de Viver            ActivitatsDeFormacio   
2023            Centre Cívic Baró de Viver  InscripcionsActivitatsFormacio   
...                                    ...                             ...   
2023              Centre Cultural Albareda  InscripcionsActivitatsFormacio   
2023              Centre Cultural Albareda                 Metres_quadrats   
2023  Centre Cultural la Farinera del Clot            ActivitatsDeFormacio   
2023  Centre Cultural la Farinera del Clot  InscripcionsActivitatsFormacio   
2023  Centre Cultural la Farinera del Clot                 Metres_quadrats   

      Valor Notes_Dades Notes_Equipament  Codi_Districte   Nom_Districte  \
Any                                                                        
2023    232         NaN              NaN               1    Ciutat Vella   
2023   2867         NaN              NaN               1    Ciutat Vella   
2023   1431         NaN              NaN               1    Ciutat Vella   
2023    137         NaN              NaN               9     Sant Andreu   
2023   1875         NaN              NaN               9     Sant Andreu   
...     ...         ...              ...             ...             ...   
2023   1297         NaN              NaN               3  Sants-Montjuïc   
2023   1473         NaN              NaN               3  Sants-Montjuïc   
2023    120         NaN              NaN              10      Sant Martí   
2023   1352         NaN              NaN              10      Sant Martí   
2023   2634         NaN              NaN              10      Sant Martí   

      Codi_Barri       Nom_Barri              Titularitat    TipusGeneral  \
Any                                                                         
2023           3  la Barceloneta  Ajuntament de Barcelona  Centres cívics   
2023           3  la Barceloneta  Ajuntament de Barcelona  Centres cívics   
2023           3  la Barceloneta  Ajuntament de Barcelona  Centres cívics   
2023          58   Baró de Viver  Ajuntament de Barcelona  Centres cívics   
2023          58   Baró de Viver  Ajuntament de Barcelona  Centres cívics   
...          ...             ...                      ...             ...   
2023          11    el Poble Sec  Ajuntament de Barcelona  Centres cívics   
2023          11    el Poble Sec  Ajuntament de Barcelona  Centres cívics   
2023          65         el Clot  Ajuntament de Barcelona  Centres cívics   
2023          65         el Clot  Ajuntament de Barcelona  Centres cívics   
2023          65         el Clot  Ajuntament de Barcelona  Centres cívics   

     TipusEquipament                        Ambit    Latitud  Longitud  
Any                                                                     
2023  Centres cívics  Multidisciplinaris i altres  41.380269  2.192301  
2023  Centres cívics  Multidisciplinaris i altres  41.380269  2.192301  
2023  Centres cívics  Multidisciplinaris i altres  41.380269  2.192301  
2023  Centres cívics  Multidisciplinaris i altres  41.448392  2.199921  
2023  Centres cívics  Multidisciplinaris i altres  41.448392  2.199921  
...              ...                          ...        ...       ...  
2023  Centres cívics  Multidisciplinaris i altres  41.374000  2.173269  
2023  Centres cívics  Multidisciplinaris i altres  41.374000  2.173269  
2023  Centres cívics  Multidisciplinaris i altres  41.404914  2.188443  
2023  Centres cívics  Multidisciplinaris i altres  41.404914  2.188443  
2023  Centres cívics  Multidisciplinaris i altres  41.404914  2.188443  

[153 rows x 15 columns]

### CORRECCIÓN DE NOMBRES DE VARIABLES

**Corrección automática**

In [24]:
df = clean_names(df) # Limpieza de nombres de columnas de forma automática
df

equipament                       indicador  \
Any                                                                          
2023              Centre Cívic Barceloneta            ActivitatsDeFormacio   
2023              Centre Cívic Barceloneta  InscripcionsActivitatsFormacio   
2023              Centre Cívic Barceloneta                 Metres_quadrats   
2023            Centre Cívic Baró de Viver            ActivitatsDeFormacio   
2023            Centre Cívic Baró de Viver  InscripcionsActivitatsFormacio   
...                                    ...                             ...   
2023              Centre Cultural Albareda  InscripcionsActivitatsFormacio   
2023              Centre Cultural Albareda                 Metres_quadrats   
2023  Centre Cultural la Farinera del Clot            ActivitatsDeFormacio   
2023  Centre Cultural la Farinera del Clot  InscripcionsActivitatsFormacio   
2023  Centre Cultural la Farinera del Clot                 Metres_quadrats   

      valor notes_dades notes_equipament  codi_districte   nom_districte  \
Any                                                                        
2023    232         NaN              NaN               1    Ciutat Vella   
2023   2867         NaN              NaN               1    Ciutat Vella   
2023   1431         NaN              NaN               1    Ciutat Vella   
2023    137         NaN              NaN               9     Sant Andreu   
2023   1875         NaN              NaN               9     Sant Andreu   
...     ...         ...              ...             ...             ...   
2023   1297         NaN              NaN               3  Sants-Montjuïc   
2023   1473         NaN              NaN               3  Sants-Montjuïc   
2023    120         NaN              NaN              10      Sant Martí   
2023   1352         NaN              NaN              10      Sant Martí   
2023   2634         NaN              NaN              10      Sant Martí   

      codi_barri       nom_barri              titularitat    tipusgeneral  \
Any                                                                         
2023           3  la Barceloneta  Ajuntament de Barcelona  Centres cívics   
2023           3  la Barceloneta  Ajuntament de Barcelona  Centres cívics   
2023           3  la Barceloneta  Ajuntament de Barcelona  Centres cívics   
2023          58   Baró de Viver  Ajuntament de Barcelona  Centres cívics   
2023          58   Baró de Viver  Ajuntament de Barcelona  Centres cívics   
...          ...             ...                      ...             ...   
2023          11    el Poble Sec  Ajuntament de Barcelona  Centres cívics   
2023          11    el Poble Sec  Ajuntament de Barcelona  Centres cívics   
2023          65         el Clot  Ajuntament de Barcelona  Centres cívics   
2023          65         el Clot  Ajuntament de Barcelona  Centres cívics   
2023          65         el Clot  Ajuntament de Barcelona  Centres cívics   

     tipusequipament                        ambit    latitud  longitud  
Any                                                                     
2023  Centres cívics  Multidisciplinaris i altres  41.380269  2.192301  
2023  Centres cívics  Multidisciplinaris i altres  41.380269  2.192301  
2023  Centres cívics  Multidisciplinaris i altres  41.380269  2.192301  
2023  Centres cívics  Multidisciplinaris i altres  41.448392  2.199921  
2023  Centres cívics  Multidisciplinaris i altres  41.448392  2.199921  
...              ...                          ...        ...       ...  
2023  Centres cívics  Multidisciplinaris i altres  41.374000  2.173269  
2023  Centres cívics  Multidisciplinaris i altres  41.374000  2.173269  
2023  Centres cívics  Multidisciplinaris i altres  41.404914  2.188443  
2023  Centres cívics  Multidisciplinaris i altres  41.404914  2.188443  
2023  Centres cívics  Multidisciplinaris i altres  41.404914  2.188443  

[153 rows x 15 columns]

**Corrección manual**

In [25]:
# df.rename(columns = {'fomacion':'formacion',
#                      'prestamo_hipotecario':'hipoteca',
#                      'prestamo_personal':'consumo'}, inplace=True)
# df

### VISIÓN GENERAL

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 153 entries, 2023 to 2023
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   equipament        153 non-null    object 
 1   indicador         153 non-null    object 
 2   valor             153 non-null    int64  
 3   notes_dades       9 non-null      object 
 4   notes_equipament  6 non-null      object 
 5   codi_districte    153 non-null    int64  
 6   nom_districte     153 non-null    object 
 7   codi_barri        153 non-null    int64  
 8   nom_barri         153 non-null    object 
 9   titularitat       153 non-null    object 
 10  tipusgeneral      153 non-null    object 
 11  tipusequipament   153 non-null    object 
 12  ambit             153 non-null    object 
 13  latitud           153 non-null    float64
 14  longitud          153 non-null    float64
dtypes: float64(2), int64(3), object(10)
memory usage: 19.1+ KB


### TIPOS DE DATOS

- Identificación

In [27]:
df.dtypes

equipament           object
indicador            object
valor                 int64
notes_dades          object
notes_equipament     object
codi_districte        int64
nom_districte        object
codi_barri            int64
nom_barri            object
titularitat          object
tipusgeneral         object
tipusequipament      object
ambit                object
latitud             float64
longitud            float64
dtype: object

- Corrección

In [28]:
# Convertir tipos de datos más apropiados
df = df.astype({
    # Mantener como object (campos de texto)
    'equipament': 'object',          # Nombre del equipamiento
    'indicador': 'object',          # Tipo de indicador
    'notes_dades': 'object',        # Notas sobre los datos (tiene nulos)
    'notes_equipament': 'object',   # Notas sobre equipamiento (tiene nulos)
    'nom_districte': 'object',      # Nombre del distrito
    'nom_barri': 'object',          # Nombre del barrio
    'titularitat': 'object',        # Tipo de titularidad
    'tipusgeneral': 'object',       # Tipo general
    'tipusequipament': 'object',    # Tipo de equipamiento
    'ambit': 'object',              # Ámbito
    
    # Mantener como int64 (códigos y valores enteros)
    'valor': 'int64',              # Valor numérico entero
    'codi_districte': 'int64',     # Código de distrito
    'codi_barri': 'int64',         # Código de barrio
    
    # Mantener como float64 (coordenadas geográficas)
    'latitud': 'float64',          # Coordenada latitud
    'longitud': 'float64'          # Coordenada longitud
})

df.dtypes

equipament           object
indicador            object
valor                 int64
notes_dades          object
notes_equipament     object
codi_districte        int64
nom_districte        object
codi_barri            int64
nom_barri            object
titularitat          object
tipusgeneral         object
tipusequipament      object
ambit                object
latitud             float64
longitud            float64
dtype: object

La mayoría de las columnas están completas (153 registros no nulos) excepto:
- notes_dades: solo 9 registros con datos
- notes_equipament: solo 6 registros con datos

Los tipos actuales son apropiados:
- Los códigos (codi_districte, codi_barri) como enteros
- Las coordenadas (latitud, longitud) como float
- Los campos descriptivos como object
- El campo valor como entero

### VALORES ÚNICOS

- Identificación

In [29]:
df.nunique().sort_values()

ambit                 1
tipusequipament       1
tipusgeneral          1
titularitat           1
notes_equipament      2
indicador             3
notes_dades           3
codi_districte       10
nom_districte        10
nom_barri            43
codi_barri           43
equipament           51
longitud             51
latitud              51
valor               146
dtype: int64

In [30]:
columnas_un_valor = [col for col in df.columns if df[col].nunique() == 1]
# ambit, tipusequipament, tipusgeneral, titularitat
print("Valores únicos en columnas con un solo valor:")
for col in columnas_un_valor:
    print(f"{col}: {df[col].unique()[0]}")

Valores únicos en columnas con un solo valor:
titularitat: Ajuntament de Barcelona
tipusgeneral: Centres cívics
tipusequipament: Centres cívics
ambit: Multidisciplinaris i altres


- Corrección

In [31]:
# Eliminar columnas con un único valor
columnas_vacias = [col for col in df.columns if df[col].nunique() == 1]
df = df.drop(columns=columnas_vacias)
print("Columnas eliminadas:", columnas_vacias)

Columnas eliminadas: ['titularitat', 'tipusgeneral', 'tipusequipament', 'ambit']


In [32]:
# 1. Análisis de variables geográficas
print("1. Distribución por distrito:")
print(df.groupby(['nom_districte', 'codi_districte'])['valor'].agg(['count', 'mean']))

print("\n2. Top 5 barrios por número de actividades:")
print(df.groupby('nom_barri')['valor'].count().sort_values(ascending=False).head())

# 3. Análisis de equipamientos
print("\n3. Distribución por equipamiento:")
equipamientos = df.groupby('equipament').agg({
    'valor': ['count', 'mean'],
    'latitud': 'first',
    'longitud': 'first'
}).round(4)
print(equipamientos.head())

# 4. Análisis del indicador
print("\n4. Tipos de indicadores:")
print(df.groupby('indicador')['valor'].agg(['count', 'mean', 'min', 'max']))

# 5. Notas y observaciones
if 'notes_equipament' in df.columns:
    print("\n5. Distribución de notas de equipamiento:")
    print(df['notes_equipament'].value_counts())

if 'notes_dades' in df.columns:
    print("\n6. Distribución de notas de datos:")
    print(df['notes_dades'].value_counts())

1. Distribución por distrito:
                                    count         mean
nom_districte       codi_districte                    
Ciutat Vella        1                  12  1202.583333
Eixample            2                  18  2622.722222
Gràcia              6                   6  1994.333333
Horta-Guinardó      7                  15   713.266667
Les Corts           4                  12  1504.166667
Nou Barris          8                  12  1130.416667
Sant Andreu         9                  21   994.095238
Sant Martí          10                 15  1707.133333
Sants-Montjuïc      3                  18  1595.888889
Sarrià-Sant Gervasi 5                  24   986.250000

2. Top 5 barrios por número de actividades:
nom_barri
Sant Gervasi - la Bonanova               6
Sarrià                                   6
Vallvidrera, el Tibidabo i les Planes    6
les Corts                                6
el Poble Sec                             6
Name: valor, dtype: int64

3. Distribuci

In [33]:
# 1. Pivotar los datos para tener los indicadores como columnas
df_pivot = df.pivot_table(
    index=['equipament', 'nom_districte', 'nom_barri'],
    columns='indicador',
    values='valor',
    aggfunc='first'
).reset_index()

# 2. Calcular ratios
df_pivot['inscripcions_per_activitat'] = (
    df_pivot['InscripcionsActivitatsFormacio'] / 
    df_pivot['ActivitatsDeFormacio']
).round(2)

df_pivot['activitats_per_metre'] = (
    df_pivot['ActivitatsDeFormacio'] / 
    df_pivot['Metres_quadrats'] * 100
).round(2)

# 3. Mostrar resultados
print("1. Top 5 centros por inscripciones por actividad:")
print(df_pivot.nlargest(5, 'inscripcions_per_activitat')[
    ['equipament', 'nom_districte', 'inscripcions_per_activitat']
])

print("\n2. Top 5 centros por densidad de actividades:")
print(df_pivot.nlargest(5, 'activitats_per_metre')[
    ['equipament', 'nom_districte', 'activitats_per_metre']
])

# 4. Estadísticas por distrito
stats_distrito = df_pivot.groupby('nom_districte').agg({
    'inscripcions_per_activitat': ['mean', 'min', 'max'],
    'activitats_per_metre': ['mean', 'min', 'max'],
    'Metres_quadrats': 'mean'
}).round(2)

print("\n3. Estadísticas por distrito:")
print(stats_distrito)

1. Top 5 centros por inscripciones por actividad:
indicador                          equipament nom_districte  \
44                        Centre Cívic Urgell      Eixample   
29              Centre Cívic La Casa Elizalde      Eixample   
20         Centre Cívic El Coll - La Bruguera        Gràcia   
24                    Centre Cívic Golferichs      Eixample   
23                    Centre Cívic Fort Pienc      Eixample   

indicador  inscripcions_per_activitat  
44                              23.17  
29                              19.75  
20                              18.50  
24                              17.86  
23                              16.61  

2. Top 5 centros por densidad de actividades:
indicador                                         equipament   nom_districte  \
27         Centre Cívic Josep M. Trias i Peitx - Riera Bl...       Les Corts   
28                                    Centre Cívic La Cadena  Sants-Montjuïc   
44                                       Cen

In [34]:
conclusiones = {
    'Éxito de programación': {
        'Eixample': 'Lidera en inscripciones y eficiencia',
        'Gràcia': 'Alta tasa de participación',
        'Les Corts': 'Mejor uso del espacio'
    },
    'Características espaciales': {
        'Sant Martí': {
            'espacios_grandes': 3417.40,  # m² promedio
            'menor_densidad': 4.38        # actividades/100m²
        },
        'Horta-Guinardó': {
            'espacios_pequeños': 856.80,  # m² promedio
            'densidad_media': 14.68
        }
    }
}

In [35]:
# 1. Análisis espacial más detallado
print("1. Distribución geográfica de actividades:")
geo_analysis = df_pivot.groupby(['nom_districte', 'nom_barri']).agg({
    'ActivitatsDeFormacio': 'sum',
    'InscripcionsActivitatsFormacio': 'sum',
    'Metres_quadrats': 'sum'
}).sort_values('InscripcionsActivitatsFormacio', ascending=False)
print(geo_analysis.head(10))

# 2. Correlaciones entre variables numéricas
print("\n2. Matriz de correlaciones:")
correlations = df_pivot[['ActivitatsDeFormacio', 
                        'InscripcionsActivitatsFormacio', 
                        'Metres_quadrats']].corr()
print(correlations.round(2))

# 3. Análisis por tamaño de centro
df_pivot['categoria_tamany'] = pd.qcut(df_pivot['Metres_quadrats'], 
                                     q=4, 
                                     labels=['Petit', 'Mitjà-petit', 
                                           'Mitjà-gran', 'Gran'])

print("\n3. Estadísticas por tamaño de centro:")
tamany_stats = df_pivot.groupby('categoria_tamany').agg({
    'Metres_quadrats': ['mean', 'min', 'max'],
    'ActivitatsDeFormacio': 'mean',
    'inscripcions_per_activitat': 'mean'
}).round(2)
print(tamany_stats)

# 4. Análisis temporal (si hay datos temporales)
if 'any' in df_pivot.columns or 'mes' in df_pivot.columns:
    print("\n4. Evolución temporal:")
    # Añadir análisis temporal aquí

# 5. Ratios adicionales
df_pivot['inscripcions_per_metre'] = (
    df_pivot['InscripcionsActivitatsFormacio'] / 
    df_pivot['Metres_quadrats']
).round(2)

print("\n5. Top 5 centros por eficiencia total:")
print(df_pivot.nlargest(5, 'inscripcions_per_metre')[
    ['equipament', 'nom_districte', 'inscripcions_per_metre']
])

1. Distribución geográfica de actividades:
indicador                                               ActivitatsDeFormacio  \
nom_districte       nom_barri                                                  
Eixample            la Nova Esquerra de l'Eixample                       795   
                    la Dreta de l'Eixample                               598   
Sants-Montjuïc      Sants                                               1099   
Gràcia              el Camp d'en Grassot i Gràcia Nova                   491   
Les Corts           la Maternitat i Sant Ramon                           777   
Sant Andreu         Navas                                                399   
Sarrià-Sant Gervasi el Putxet i el Farró                                 351   
Les Corts           les Corts                                            430   
Eixample            la Sagrada Família                                   325   
Sarrià-Sant Gervasi Sant Gervasi - la Bonanova                           247 

C:\Users\Usuario\AppData\Local\Temp\ipykernel_15168\2292553290.py:24: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  tamany_stats = df_pivot.groupby('categoria_tamany').agg({


In [36]:
categorias_tamany = {
    'Petit': {
        'media_m2': 660,
        'actividades': 155.92,
        'ratio_inscripcion': 11.48
    },
    'Mitjà-petit': {
        'media_m2': 1035.21,
        'actividades': 193.50,
        'ratio_inscripcion': 12.32
    },
    'Mitjà-gran': {
        'media_m2': 1471.27,
        'actividades': 160.36,
        'ratio_inscripcion': 12.88
    },
    'Gran': {
        'media_m2': 2761.15,
        'actividades': 299.54,
        'ratio_inscripcion': 11.45
    }
}

# El Eixample domina en eficiencia y actividad
# El tamaño del centro no determina su éxito
# Los centros medianos tienen mejor ratio de inscripciones
# Alta correlación entre actividades e inscripciones
# El espacio físico no es determinante para el éxito


### DUPLICADOS

- Identificación

In [37]:
df.duplicated().sum()

np.int64(0)

- Corrección

In [38]:
# df.drop_duplicates(inplace = True)

### SEPARAR NUMÉRICAS Y CATEGÓRICAS

**Categóricas**

In [39]:
cat = df.select_dtypes(exclude = 'number').copy()

**Numéricas**

In [40]:
num = df.select_dtypes(include='number').copy()

### GESTIÓN DE CATEGÓRICAS

#### NULOS

Identificación

In [41]:
cat.isna().sum().sort_values(ascending = False)

notes_equipament    147
notes_dades         144
indicador             0
equipament            0
nom_districte         0
nom_barri             0
dtype: int64

Corrección

- Eliminar la variable

In [42]:
# Opción 1: Eliminar estas columnas por tener demasiados nulos
cat = cat.drop(columns=['notes_equipament', 'notes_dades'])

# # Opción 2: Mantenerlas pero imputar con un valor específico
# cat['notes_equipament'] = cat['notes_equipament'].fillna('Sin notas')
# cat['notes_dades'] = cat['notes_dades'].fillna('Sin notas')

# Opción 3: Mantenerlas tal cual si la información, aunque escasa, es valiosa

- Eliminar registros

*Cuidado: Si se eliminan registros hay que eliminarlos tanto de cat como de num*

In [43]:
# criterio = 3
# cat.dropna(thresh=criterio, inplace=True)
# num.dropna(thresh=criterio, inplace=True)

- Imputar por un valor

Variables a imputar

In [44]:
# var_imputar_valor = ['resultado_campana_anterior','formacion']

Valor por el que imputar

In [45]:
# valor = 'NULO'

- Imputar

In [46]:
# cat[var_imputar_valor] = cat[var_imputar_valor].fillna(valor)

- Imputar por la moda

Variables a imputar

In [47]:
# var_imputar_moda = ['resultado_campana_anterior','formacion']

Funcion para imputar con la moda

In [48]:
# def imputar_moda(variable):
#     return(variable.fillna(variable.mode()[0]))

Imputar con la moda

In [49]:
# cat[var_imputar_moda] = cat[var_imputar_moda].apply(imputar_moda)

### ATÍPICOS

- Categorías poco frecuentes

Función para agrupar categorías raras en 'OTROS'

In [50]:
# def agrupar_cat_raras(variable, criterio = 0.05):
#     # Calcula las frecuencias
#     frecuencias = variable.value_counts(normalize=True)
#     # Identifica las que están por debajo del criterio
#     temp = [cada for cada in frecuencias.loc[frecuencias < criterio].index.values]
#     # Las recodifica en 'OTROS'
#     temp2 = np.where(variable.isin(temp),'OTROS',variable)
#     # Devuelve el resultado
#     return(temp2)

Variables a imputar

In [51]:
# var_imputar_moda = ['trabajo','estado_civil']

Aplicar

In [52]:
# for variable in var_imputar_moda:
#     cat[variable] = agrupar_cat_raras(cat[variable],criterio = 0.02)

### GESTIÓN DE NUMÉRICAS

#### NULOS

Identificación

In [53]:
num.isna().sum().sort_values(ascending = False)

valor             0
codi_districte    0
codi_barri        0
latitud           0
longitud          0
dtype: int64

Corrección

- Eliminar la variable

In [54]:
# num.drop(columns = 'num_contactos_otras_campanas',inplace = True)

- Eliminar registros

*Cuidado: Si se eliminan registros hay que eliminarlos tanto de cat como de num*

In [55]:
# criterio = 3
# cat.dropna(thresh=criterio, inplace=True)
# num.dropna(thresh=criterio, inplace=True)

- Imputar por la media

Variables a imputar

In [56]:
# var_imputar_media = ['edad','num_contactos_esta_campana']

Funcion para imputar con la media

In [57]:
# def imputar_media(variable):
#     if pd.api.types.is_integer_dtype(variable):
#         return(variable.fillna(int(variable.mean())))
#     else:
#         return(variable.fillna(variable.mean()))

- Imputar con la media

In [58]:
# num[var_imputar_media] = num[var_imputar_media].apply(imputar_media)

- Imputar por la mediana

Variables a imputar

In [59]:
# var_imputar_mediana = ['edad','num_contactos_esta_campana']

Funcion para imputar con la mediana

In [60]:
# def imputar_mediana(variable):
#     if pd.api.types.is_integer_dtype(variable):
#         return(variable.fillna(int(variable.median())))
#     else:
#         return(variable.fillna(variable.median()))

- Imputar con la mediana

In [61]:
# num[var_imputar_mediana] = num[var_imputar_mediana].apply(imputar_mediana)

### ATÍPICOS

**Por desviación típica**

- Identificación

Definir el número de desviaciones típicas

In [62]:
# num_desv_tip = 4

Función que devuelve los índices de los atípicos

In [63]:
# def atipicos_desv_tip(variable, num_desv_tip = 4):
#     #sacamos los nulos por ahora
#     variable = variable.dropna()
#     #calculamos los límites
#     media = np.mean(variable)
#     sd = np.std(variable)
#     umbral = sd * num_desv_tip
#     lim_inf = media - umbral
#     lim_sup = media + umbral
#     #encontramos los índices de los que están fuera de los límites
#     indices = [indice for indice,valor in variable.items() if valor < lim_inf or valor > lim_sup]
#     return(indices)

Función que cuenta el número de atípicos

In [64]:
# def conteo_atipicos(df,variable, num_desv_tip = 4):
#     atipicos = atipicos_desv_tip(df[variable], num_desv_tip)
#     return(df.loc[atipicos,variable].value_counts())

Variables a aplicar

In [65]:
# var_atipicos_dt = ['edad','num_contactos_esta_campana']

Aplicar

In [66]:
# for variable in var_atipicos_dt:
#     print('\n' + variable + ':\n')
#     print(conteo_atipicos(num,variable,num_desv_tip))

- Corrección

Eliminar los registros que tengan al menos un atípico en alguna de las variables indicadas.

*Cuidado: Si se eliminan registros hay que eliminarlos tanto de cat como de num*

Variables a aplicar

In [67]:
# var_atipicos_dt = ['edad','num_contactos_esta_campana']

Aplicar

In [68]:
# for variable in var_atipicos_dt:
#     cat.drop(atipicos_desv_tip(num[variable], num_desv_tip),axis = 0, inplace = True)
#     num.drop(atipicos_desv_tip(num[variable], num_desv_tip),axis = 0, inplace = True)

**Por winsorización manual para una variable** 

- Identificación

Variable a analizar

In [69]:
# variable = 'variacion_tasa_empleo'

Comprobar el mínimo y máximo actual

In [70]:
# num[variable].describe().loc[['min','max']]

- Corrección

Definir el mínimo y máximo tolerable.

In [71]:
# minimo = -20
# maximo = 10

Aplicar

In [72]:
# num[variable].clip(minimo,maximo, inplace=True)

**Por winsorización automática para múltiples variables**

Variables a winsorizar

In [73]:
# var_winsorizar = ['edad','num_contactos_esta_campana']

Definir los percentiles mínimos y máximos.

In [74]:
# p_min = 0.01
# p_max = 0.99

Aplicar

In [75]:
# for variable in var_winsorizar:
#     num[variable].clip(lower = num[variable].quantile(p_min), 
#                        upper = num[variable].quantile(p_max), 
#                        inplace = True)

### GUARDAR DATASETS TRAS CALIDAD DE DATOS

Guarda los avances en df, cat y num con un sufijo para poder guardar o recuperar avances del proyecto.

En formato pickle para no perder las modificaciones de metadatos.

- Definir los nombres de los archivos

In [76]:
ruta_trabajo = ruta_proyecto + '/02_Datos/03_Trabajo/' + '2023_centres_civics_activitats_formacio.pickle'
ruta_cat = ruta_proyecto + '/02_Datos/03_Trabajo/' + 'cat_2023_centres_civics_activitats_formacio.pickle'
ruta_num = ruta_proyecto + '/02_Datos/03_Trabajo/' + 'num_2023_centres_civics_activitats_formacio.pickle'

- Guardar los archivos

In [77]:
df.to_pickle(ruta_trabajo)
cat.to_pickle(ruta_cat)
num.to_pickle(ruta_num)